File: Problem Sets/SinaNozhatie-credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool validity(long number);
bool checksum(long number);
int find_lenth(long number);
void choose_bank(long number);

int main(void)
{
    long number;
    do
    {
        number = get_long("Enter Card Number:");

    }
    while (number < 0); // Card number must be Greater than 0

    if (validity(number) == true) //checks validity of card
    {

        choose_bank(number); //Printing the bank name if it was valid

    }
    else
    {
        printf("INVALID\n");
    }
}

bool validity(long number)
{
    int lenth = find_lenth(number); //check the lenth

    if ((lenth == 13 || lenth == 15 || lenth == 16) && (checksum(number) == true))
    {
        return true;
    }
    else
    {
        return false;
    }


}

int find_lenth(long number)
{
    int lenth = 0;

    while (number != 0) //this loop remove the last digit untill the number reach to 0
    {
        lenth++;
        number /= 10; //remove the last Digit
    }
    return lenth;

}

bool checksum(long number)
{
    int sum1 = 0, sum2 = 0, sum;
    int i = 0;

    while (number != 0)
    {
        if (i % 2 != 0) //if it's the second last digit
        {   
            
            int n = (2 * (number % 10)); // multiply the last digit by 2
            

            if (n > 9)
            {  
                sum1 += n % 10 + n / 10;  //if the number was 12 do 1 + 2
            }
            else
            {
                sum1 += n;
                            

            }

        }
        else
        {
            sum2 += number % 10;
        }
        number /= 10;//remove the last Digit
        i++;
    }

    sum = sum1 + sum2; //sum of digits must = 20
        
    if (sum % 10 == 0)// the last digit must = 0
    {
        return true;
    }
    else
    {
        return false;
    }
}

void choose_bank(long number)
{


    if (number / 10000000000000 == 34 || number / 10000000000000 == 37) //if the first 2 digits is 34 or 37
    {
        printf("AMEX\n");
    }

    else if (number / 100000000000000  == 51 || number / 100000000000000  == 52 || number / 100000000000000  == 53
             || number / 100000000000000  == 54 || number / 100000000000000  == 55)
    {
        printf("MASTERCARD\n");
    }

    else if (number / 1000000000000  == 4 || number / 1000000000000000 == 4) //if the first digit is 4
    {
        printf("VISA\n");
    }
    
    else
    {
        printf("INVALID\n");
    }

}

File: Problem Sets/SinaNozhatie-Python-credit/credit.py
# importing libraries

from cs50 import get_int
import re


def main():
    # infinite loop if user enter letter or negative int
    while True:
        cardnumber = get_int("card number:")
        if cardnumber > 0:
            break

    if validity(cardnumber) == True:
        choose_bank(cardnumber)

    else:
        print("INVALID")


def validity(cardnumber):
    # check the lenth of the cardnumber
    lenth = len(str(cardnumber))

    # if the checksum and lenth both = True then the function returns True
    if (lenth == 13 or lenth == 15 or lenth == 16) and (checksum(cardnumber) == True):
        return True

    else:
        return False


def checksum(cardnumber):
    sum1 = 0
    sum2 = 0
    i = 0

    while cardnumber != 0:
        if i % 2 == 0:  # if it's the last digit add it to sum2
            sum2 += cardnumber % 10

        else:
            n = 2 * (cardnumber % 10)  # multiply the second last digit by 2

            if n > 9:
                sum1 += (
                    n % 10 + n / 10
                )  # if the number was 12 do 1 + 2 then add to sum1

            else:
                sum1 += n
        # int(x) for removing the decimals
        sum1 = int(sum1)
        sum2 = int(sum2)
        cardnumber = cardnumber / 10  # remove the last Digit
        cardnumber = int(cardnumber)
        i += 1

    sum = sum1 + sum2

    # the last digit of sum must = 0

    if sum % 10 == 0:
        return True

    else:
        return False


def choose_bank(cardnumber):
    # sorting a list of RegEx
    pattern = ["^34", "^37", "^51", "^52", "^53", "^54", "^55", "^4"]
    for i in range(8):
        # check each RegEx in pattern with cardnumber
        result = re.match(pattern[i], str(cardnumber))
        if result:  # if they match the loop will stop
            break
        else:
            i += 1  # to print INVALID

    # i is the number of RegEx in list
    if i == 0 or i == 1:
        print("AMEX")

    elif i == 2 or i == 3 or i == 4 or i == 5 or i == 6:
        print("MASTERCARD")

    elif i == 7:
        print("VISA")

    else:
        print("INVALID")


main()
